Issue # 767 Display Local Time Zones for Revisions in Search Results and Selected Revisions

hibaa03 opened the pull request:
This pull request implements the following changes close #767 :

**Display Local Time Zones:**
Added functionality to display the local time zone alongside each revision timestamp. It shows me time in my local timezone (GMT+5) as in the screenshot below.
![Image 17-10-2024 at 11 27 PM](https://github.com/user-attachments/assets/bdf0e77b-6d61-43b1-ad39-d9041ccf3c33)




**UTC Timestamp Tooltip:**
Introduced a tooltip that shows the UTC timestamp when hovering over the local time. 
![Image 17-10-2024 at 11 27 PM (1)](https://github.com/user-attachments/assets/313c06cd-4bd8-46b6-aeb8-41b0e224d702)


**Time Zone in Selected Revision:**
Ensured the local time zone is also displayed for any selected revision, along with the UTC on hover through a tooltip.
![Image 17-10-2024 at 11 27 PM (2)](https://github.com/user-attachments/assets/29cb66e9-4962-4396-8179-6799c688f883)

**Compactness:**
Displayed time and time zone in a more digestible format. Created a separate DateTime component to handle both the GMT display and UTC tooltip functionality, simplifying the time display logic.

syedbarimanjan commented on src/__tests__/Search/SearchResultsList.test.tsx:
You have accidentally removed the comments from this file.

hibaa03 replied:
thanks for pointing it out, fixed it.

syedbarimanjan replied:
You have put some comments on wrong lines.

hibaa03 replied:
could you please point me to it?

hibaa03 replied:
sorry nevermind. working on it

syedbarimanjan commented:
Add a ```close #bug-number``` to your pr description so that if the pr gets merged the issue will close automatically.

Carla-Moz commented:
Great effort, Hiba, however, the design requires a more compact version of the timezones. See below. 

![image](https://github.com/user-attachments/assets/129383d0-078b-4a33-bfba-6f7d05aefe1a)

Your version also breaks the style for my time zone:

![image](https://github.com/user-attachments/assets/b412c2bb-eeaa-4b65-9364-9ae4681cd1dc)

hibaa03 commented:
Hi @Carla-Moz, I appreciate your feedback. I have made the changes accordingly, could you please review it when possible?

julienw commented on package.json:
```diff
@@ -24,13 +24,14 @@
     "json-e": "^4.7.0",
     "material-ui-popup-state": "^5.1.2",
     "moize": "^6.1.6",
+    "moment-timezone": "^0.5.46",
```
We don't really want to add `moment` as we already have `dayjs` :-) `moment` is a huge library.

Also nowadays the web platform already has this functionality, you can read about it on MDN => https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString
It's possible we can replace `dayjs` in the process too.

hibaa03 replied:
Initially I used day.js with timezone plugin to handle time zones, but the output format is limited to either IANA timezones (America/New_York) or the GMT offset (GMT+/-X), as day.js doesn’t provide abbreviations like PST or EST, so I added moment (as to make it look like the screenshot Carla has sent above, you can also see how it was before when I used dayjs (in the second screenshot she shared)). Let me know how you’d prefer me to proceed!

hibaa03 replied:
![Image 11-10-2024 at 7 49 PM](https://github.com/user-attachments/assets/dcfd018f-2c17-49e2-a105-517711da20ca)
This is also what I can get with day.js. Would this be better and compact (compared to IANA timezones) so I can make the changes accordingly?

hibaa03 replied:
I have made the changes! Please review them when possible

julienw replied:
Can you remove this change from package.json and run npm i again?

julienw commented on src/__tests__/CompareResults/ResultsView.test.tsx:
```diff
@@ -16,6 +16,8 @@ import {
   FetchMockSandbox,
 } from '../utils/test-utils';
 
+jest.setTimeout(60000);
```
you probably don't need this change :-) right? Can you explain why you added these?

hibaa03 replied:
I added the jest.setTimeout(60000) to prevent timeout issues during tests as they were failing because of it. The only way I could solve it and have all tests passed

julienw reviewed (changes requested):
Adding `moment` isn't great for us because it's a huge library. Instead you can use either `dayjs` that we already have, or directly the web platform API `toLocaleString` that I mention in my comment below.

Both options would work for me, feel free to pick one.
dayjs documentation about timezones is in https://day.js.org/docs/en/plugin/timezone.

julienw commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -66,7 +72,19 @@ function SearchResultsListItem({
+
+  const itemDate = dayjs.unix(item.push_timestamp);
+
+  const date = itemDate.format('MM/DD/YY');
```
The screenshot Carla shared in https://github.com/mozilla/perfcompare/issues/767 shows the date and time using the USA notation: MM/DD/YY. But this is confusing to all other users. Indeed in most other parts of the world, we use DD/MM/YY. Therefore for some dates it's confusing. For example 01/10/24 could be either January 10 or October 1st.

Therefore my recommendation would be to use `MMM` instead of `MM`, so that the first letters of the month is used. Also it would be better to use a single digit for the day when possible, so use `D` instead of `DD`.
Then I believe the format would look like `MMM D, YY`, but please check if this looks right.

This wasn't mentioned in the issue, sorry about that.

julienw commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -66,7 +72,19 @@ function SearchResultsListItem({
+  const itemDate = dayjs.unix(item.push_timestamp);
+
+  const date = itemDate.format('MM/DD/YY');
+  const localTime = itemDate.format('HH:mm');
```
similarly here, it would be better to have just 1 digit when possible, therefore you can use `H:m`.

julienw commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -66,7 +72,19 @@ function SearchResultsListItem({
+  const localTime = itemDate.format('HH:mm');
+
+  // convert timezone to GMT
+  const gmtOffset = `GMT${itemDate
+    .format('Z')
+    .replace(':00', '')
+    .replace(/^([+-])0?/, '$1')}`;
```
I don't like this a lot, but I don't see a better solution using only dayjs :/

julienw commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -8,14 +8,21 @@ import ListItemButton from '@mui/material/ListItemButton';
 
+// day.js with timezone support
+dayjs.extend(utc);
+dayjs.extend(timezone);
```
I think you don't use the timezone plugin anymore, can you please remove this line?

julienw commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -66,7 +72,19 @@ function SearchResultsListItem({
+    .replace(':00', '')
+    .replace(/^([+-])0?/, '$1')}`;
+
+  const utcTime = itemDate.utc().format('MM/DD/YY HH:mm [UTC]');
```
same comments as above :)

julienw commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -125,7 +142,19 @@ function SearchResultsListItem({
+                    <Tooltip title={`${utcTime}`}>
+                      <span>
+                        {date}{' '}
+                        <span style={{ fontWeight: 'bold' }}>
```
not sure why you changed the style, I think it's not needed. Then you can remove the spans as well

julienw commented on src/components/Search/SearchResultsListItem.tsx:
```diff
@@ -125,7 +142,19 @@ function SearchResultsListItem({
-                    {String(dayjs(itemDate).format('MM/DD/YY HH:mm'))}
+
+                    {/* Tooltip with UTC time */}
+                    <Tooltip title={`${utcTime}`}>
```
it would be good to extract this functionality to a separate react component, so that you can use it both here and in SelectedRevisionItem. Can you please do that?

julienw commented on package.json:
```diff
@@ -24,13 +24,14 @@
     "react-dom": "^18.2.0",
     "react-redux": "^8.0.2",
-    "react-router-dom": "^6.18.0",
+    "react-router-dom": "^6.26.2",
```
can you revert this change pleaase ? :-)

julienw reviewed (changes requested):
Thanks Hiba, this is better!
here are some more comments :-) thanks again

hibaa03 commented:
@julienw thank you for the thorough feedback, I have made the changes accordingly. I reverted the changes made before, fixed the date and time format as you corrected, and created a separate component DateTime. 
Please give it a look when you are able to and let me know if there is anything else to change! :)

julienw commented:
Hey @hibaa03 , I merged the other PR that does the same thing (https://github.com/mozilla/perfcompare/pull/768). The main difference is that that PR used an API from the web platform instead of the library, which should be more performant and also takes less space in the bundle.
I'm closing your PR now, but this was really good work, congrats again!
